recipe/migrations/numpy2.yaml

```diff
@@ -0,0 +1,75 @@
+__migrator:
+  build_number: 1
+  kind: version
+  commit_message: |
+    Rebuild for numpy 2.0
+    
+    TL;DR: The way we build against numpy has changed as of numpy 2.0. This bot
+    PR has updated the recipe to account for the changes (see below for details).
+    The numpy 2.0 package itself is currently only available from a special release
+    channel (`conda-forge/label/numpy_rc`) and will not be available on the main
+    `conda-forge` channel until the release of numpy 2.0 GA.
+    
+    The biggest change is that we no longer need to use the oldest available numpy
+    version at build time in order to support old numpy version at runtime - numpy
+    will by default use a compatible ABI for the oldest still-supported numpy versions.
+    
+    Additionally, we no longer need to use `{{ pin_compatible("numpy") }}` as a
+    run requirement - this has been handled for more than two years now by a
+    run-export on the numpy package itself. The migrator will therefore remove
+    any occurrences of this.
+    
+    However, by default, building against numpy 2.0 will assume that the package
+    is compatible with numpy 2.0, which is not necessarily the case. You should
+    check that the upstream package explicitly supports numpy 2.0, otherwise you
+    need to add a `- numpy <2` run requirement until that happens (check numpy
+    issue 26191 for an overview of the most important packages).
+    
```

Comment:
```suggestion
    
    Additional would check what upstreams NumPy version constraints are.
    If they set `NPY_TARGET_VERSION` to a different value or express a
    different lower bound for `numpy`, please incorporate that into the package's
    `numpy` constraints.
    
```

Reply:
@rgommers made the case that numpy doesn't want anyone playing around with `NPY_TARGET_VERSION`, up-to-and-including not giving it too much visibility (so people are tempted) by using it in conda-forge. I think this aspect is a level of detail that's much too high for such a large-scale migration.

Reply:
Certainly in most cases developers probably don't need to set this

Unfortunately some appear to be setting it anyway. Ran into this in `numexpr` just the other day ( https://github.com/pydata/numexpr/pull/478#discussion_r1585553958 ). Admittedly maybe `numexpr` doesn't need it either, but until they drop it we do need to be aware of the issues from `NPY_TARGET_VERSION`

So maintainers may need to check if the particular library is doing something unusual and act accordingly. Unless we help provide tooling around doing that. Or do the whack-a-mole approach of removing from every library we find setting it

Reply:
I am still strongly -1 on mentioning `NPY_TARGET_VERSION` at all. The `numexpr` usage is unnecessary (and I assume it will be removed after your comment there), and clearly not a problem for the conda-forge `numexpr` package since all they tried to do was target the lowest available numpy (1.19, so will always work).

That said, the rest of this proposed change is useful probably. Basically: check constraints in `pyproject.toml` and ensure that `meta.yaml` is consistent with those. This isn't unique to `numpy`, it applies to all dependencies, but it's still a good reminder since it's easy to forget.

Reply:
I updated the To-Dos in https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5790/commits/8dc036b3ce3bad5d835b9a5f42478dc1527c3222 to account for that - PTAL. In the meantime, resolving this thread.